lightning-block-sync/src/init.rs

```diff
@@ -0,0 +1,267 @@
+use crate::{BlockSource, BlockSourceResult, Cache, ChainListener, ChainNotifier};
+use crate::poll::{ChainPoller, Validate, ValidatedBlockHeader};
+
+use bitcoin::blockdata::block::{Block, BlockHeader};
+use bitcoin::hash_types::BlockHash;
+use bitcoin::network::constants::Network;
+
+/// Performs a one-time sync of chain listeners using a single *trusted* block source, bringing each
+/// listener's view of the chain from its paired block hash to `block_source`'s best chain tip.
+///
+/// Upon success, the returned header can be used to initialize [`SpvClient`]. In the case of
+/// failure, each listener may be left at a different block hash than the one it was originally
+/// paired with.
+///
+/// Useful during startup to bring the [`ChannelManager`] and each [`ChannelMonitor`] in sync before
+/// switching to [`SpvClient`].
+///
```

Comment:
Excellent idea! This may get a little verbose as I'll need to implement `ChainListener` a few times. Do you think it is worth adding these implementations to the crate? I left them off from the draft but I could see them being useful to others since they would be commonly used adapters.

Reply:
Yea, I presume there's little point not having them given the crate already depends on `lightning`. Feel free to do it in the next PR and add docs then.

Reply:
FWIW, turns out the crate dependency was only needed for these implementations. It was an oversight to include it in this PR.

Reply:
Ah. I mean I guess we could feature-gate it if we want in the future, but for now I see little reason to.

Reply:
Took a stab at writing an example, which turned out to be a good exercise in seeing how suitable the interface works in practice. See 1d2f5b3d637613904f39a490ba1fdfb34e5ba62a.

I ended up changing `ChainListener`'s methods to take `&self` instead of `&mut self`. This was probably not entirely necessary but did make the code a little cleaner as otherwise the mutable borrows needed to be properly scoped. Definitely open to changing this if you have any strong opinions.

This did require implementing `ChainListener` for `RefCell<ChannelMonitor>` since it needs to call methods that take `&mut self`. I don't have a firm grasp of when trait methods should use borrows vs mutable borrows, but this seems to give some advice:

https://doc.rust-lang.org/std/cell/index.html#when-to-choose-interior-mutability

Ideally, we are at least consist in how we decide to design these interfaces that goes beyond "placates the borrow checker". :P